### ISSUE TITLE
Manual frame extraction is very slow (version 3.0.0rc14)

### ISSUE THREAD

**Issue by ChaoqunYinNeuroscience** [DeepLabCut/DeepLabCut#3264]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating System

Windows 11 Pro

### DeepLabCut version

3.0.0rc14

### What engine are you using?

pytorch

### DeepLabCut mode

single animal

### Device type

NVIDIA GeForce RTX 4080 + Intel Core i7-14700K

### Bug description 🐛

Manual frame extraction is very slow. Loading the selected frame or extracting it takes about 30 seconds per frame. Automatic frame extraction is running normally. I have tried both .avi and .mp4 formats, both at the same slow speed. 
My videos are 640*512, 80 frames/second. The video length is between 1 and 2 hours. The video size is around 4GB.
I previously used versions 2.1.5.2 and 2.3.4. There was not this issue.

### Relevant log output

```shell

```

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by deruyter92** [comment]
@ChaoqunYinNeuroscience, thanks for raising this issue. 

For manual extraction, the performance depends on how napari (or napari‑deeplabcut) reads and seeks in long videos. Note that the automatic frame extraction (using ffmpeg / OpenCV) is faster, but the mentioned 30s per video for manual frame extraction is not expected/preferred behavior. Indeed when trying to reproduce, I could observe slightly long decoding times for a long and highly-compressed video, but not close to your durations. 

At the moment, we are currently working on a fully refreshed version the napari-deeplabcut plugin, so your issue comes at the right time. To verify whether indeed there is a performance regression compared to previous versions, we would like ask for some more information:

- Which version of napari and the napari-deeplabcut plugin do you have installed? 
- And which version of OpenCV and ffmpeg?
- Are all videos on a local machine or are you using a network drive? 
- Did you compare the performance between different versions with the same video in your current setup, or is this based on an earlier experience?
- What is the video codec (e.g. H.264/AVC, H.265/HEVC, MJPEG, ProRes, etc.)?
- Could you maybe share your video with us?

**Comment by deruyter92** [comment]
By the way, maybe an easy workaround would be split the videos into separate clips. Would that work for you? Napari can handle shorter videos much faster. 

e.g. to use ffmpeg for splitting videos by time range, you could use the following command:
```
ffmpeg -i input.mp4 -ss 00:00:00 -to 00:01:00 -c copy part1.mp4
```

**Comment by ChaoqunYinNeuroscience** [comment]
Hi Jaap, thanks for your fast reply! Here are my answers to your questions:
1) napari 0.6.6 / napari-deeplabcut 0.2.1.8
2) opencv-python 4.11.0.86 / imageio-ffmpeg 0.6.0
3) The videos are local. I have tried both SSDs and regular hard drives. The speed is basically the same.
4) I tested the loading speed with exactly the same video.
5) Codec: H264 - MPEG-4 AVC (part 10) (avc1)
6) Sure thing. The files are quite big. I can share it with you over Google Drive.

Yes, I am currently cutting the videos into shorter pieces before loading. It does help. I just feel that there must be something wrong with my frame loading. ~30s per frame is abnormally slow.

**Comment by C-Achard** [comment]
Hello @ChaoqunYinNeuroscience 
We have now released a new version of the plugin (see [patch notes here](https://github.com/DeepLabCut/napari-deeplabcut/releases/tag/v0.3.0.0)), which you can install by using:
`pip install napari-deeplabcut==0.3.0.0`

I have performed some stress testing with synthetic data and did not notice heavy slowdowns; if you'd like you could try again on the new version and let us know if it performs as expected.

Alongside this, we have also [updated the documentation](https://deeplabcut.github.io/DeepLabCut/docs/gui/napari/basic_usage.html#).

Thanks,
Cyril

**Comment by ChaoqunYinNeuroscience** [comment]
Hi Cyril, thanks for letting me know! I'm giving it a shot this week. I'll let you know if it works well at my end.

**Comment by ChaoqunYinNeuroscience** [comment]
Hi Cyril, I just tested. napari-deeplabcut 0.3.0.0 is still slow, though better than the old version. There is still a noticeable delay (~10 seconds) between dragging the video progress bar and displaying the new frame.

My RAM, CPU, and GPU should be very powerful (64 GB DDR5 + Intel Core i7-14700 KF + RTX 4080 16GB).

**Comment by ChaoqunYinNeuroscience** [comment]
BTW, I noticed _a potentially serious bug_ when I ran the new napari-deeplabcut.

After extracting outlier frames and relabeling them, the "CollectedData" .csv file automatically generated by DeepLabCut contains the likelihood values from the "machinelabels" file. However, the original labels used for initial training do not have this parameter. As a result, this mismatch in the number of columns prevents me from generating a new training dataset. Moreover, because these likelihood parameters are also saved in the "CollectedData" .h5 file, I cannot simply fix the issue by deleting the columns.

**Comment by C-Achard** [comment]
> Hi Cyril, I just tested. napari-deeplabcut 0.3.0.0 is still slow, though better than the old version. There is still a noticeable delay (~10 seconds) between dragging the video progress bar and displaying the new frame.
> 
> My RAM, CPU, and GPU should be very powerful (64 GB DDR5 + Intel Core i7-14700 KF + RTX 4080 16GB).

Most likely the problem is not hardware; I remember this previous issue (https://github.com/DeepLabCut/napari-deeplabcut/issues/107), which involved the video reader.

It could be worth trying decord2 rather than opencv or pims, I will check if I can quickly add support.

**Comment by C-Achard** [comment]
> BTW, I noticed _a potentially serious bug_ when I ran the new napari-deeplabcut.
> 
> After extracting outlier frames and relabeling them, the "CollectedData" .csv file automatically generated by DeepLabCut contains the likelihood values from the "machinelabels" file. However, the original labels used for initial training do not have this parameter. As a result, this mismatch in the number of columns prevents me from generating a new training dataset. Moreover, because these likelihood parameters are also saved in the "CollectedData" .h5 file, I cannot simply fix the issue by deleting the columns.

That is correct, this is reported in #3319, and fixed in [version 0.3.0.1 of napari-deeplabcut](https://github.com/DeepLabCut/napari-deeplabcut/releases/tag/v0.3.0.1), released just now.
#3323 is not release yet however, so DLC will still fail for affected projects for now.

There are two options to fix this for affected projects:

- Easiest: update napari-deeplabcut, open each annotated (labeled-data) folder in napari and save, even without changes
- Quickest (if you're okay with using a script to edit your data):
  - Backup your labeled-data folder
  - Edit the paths and run in preview mode:
```python
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path(r"C:\path\to\your\dlc-project")  # folder containing config.yaml
LABELED_DATA = PROJECT_ROOT / "labeled-data"
PREVIEW_ONLY = True # change to False to edit the files

def drop_likelihood_from_h5(h5_path: Path, dry_run: bool = False):
    try:
        df = pd.read_hdf(h5_path)
    except Exception as e:
        print(f"[ERROR] {h5_path}: could not read H5 ({e})")
        return "error", 0

    if not isinstance(df.columns, pd.MultiIndex):
        print(f"[SKIP]  {h5_path}: columns are not a MultiIndex")
        return "skipped", 0

    coord_level = "coords" if "coords" in df.columns.names else df.columns.names[-1]
    coords = df.columns.get_level_values(coord_level)

    if "likelihood" not in coords:
        print(f"[OK]    {h5_path}: no likelihood columns found")
        return "unchanged", 0

    if "x" not in coords or "y" not in coords:
        print(f"[SKIP]  {h5_path}: coords do not include both x and y")
        return "skipped", 0

    keep_mask = coords.isin(["x", "y"])
    removed = int((~keep_mask).sum())
    df_xy = df.loc[:, keep_mask]

    print(f"[FIX]   {h5_path}: removing {removed} likelihood column(s)")

    if not dry_run:
        df_xy.to_hdf(h5_path, key="df_with_missing", mode="w")
        df_xy.to_csv(h5_path.with_suffix(".csv"))

    return "updated", removed

n_updated = 0
n_unchanged = 0
n_skipped = 0
n_errors = 0
total_removed = 0

for h5_path in sorted(LABELED_DATA.rglob("CollectedData_*.h5")):
    if "machine" in h5_path.name.lower():
        print(f"[SKIP]  {h5_path}: looks like a machine-label file")
        n_skipped += 1
        continue

    status, removed = drop_likelihood_from_h5(h5_path, dry_run=PREVIEW_ONLY)
    total_removed += removed

    if status == "updated":
        n_updated += 1
    elif status == "unchanged":
        n_unchanged += 1
    elif status == "skipped":
        n_skipped += 1
    elif status == "error":
        n_errors += 1

print("\n=== Summary ===")
if PREVIEW_ONLY:
    print("PREVIEW MODE, NO CHANGES WILL BE MADE")
print(f"Updated files:    {n_updated}")
print(f"Unchanged files:  {n_unchanged}")
print(f"Skipped files:    {n_skipped}")
print(f"Errors:           {n_errors}")
print(f"Columns removed:  {total_removed}") 
```
- Check that only the correct files are affected
- Set `PREVIEW_ONLY` to False and run again

This will also fix the stray column, but please make sure to backup.

Sorry for the disruption, and thanks for letting us know!

Hope this helps,
Cyril

**Comment by C-Achard** [comment]
I have opened https://github.com/DeepLabCut/napari-deeplabcut/issues/209 to track this anew in napari-DLC, will follow up with a PR soon! Sorry again for the delay

**Comment by ChaoqunYinNeuroscience** [comment]
Okay, I'll try to load a different video reader. What was the reader used by version 2.1.5.2 and 2.3.4? They worked perfectly for me.

Also, thank you for providing the pandas code. I tried to write one myself but eventually went back to napari-deeplabcut 0.2.1.8

**Comment by C-Achard** [comment]
Hello again @ChaoqunYinNeuroscience,

I have now made some improvements in https://github.com/DeepLabCut/napari-deeplabcut/issues/209; the problem is that performance strongly depends on the video encoding/GOP and therefore results can greatly vary across containers/codecs/settings, which is why I initially missed it with my previous synthetic data test, though it had a large number of frames as well.
I now tried with a much larger GOP and also saw similar delays as you mentioned; the new version results in a massive speedup in that setting, though for very large videos delays still occur on chunk recompute.
Still, it should be far more responsive overall; navigating in adjacent frames should be much faster too.

This can be installed with:
`pip install "git+https://github.com/DeepLabCut/napari-deeplabcut.git@refs/pull/215/head"`
**(Note that this is 0.3+ based, since you downgraded I would recommend testing this dev branch of napari-deeplabcut on its own in a separate env to preserve your current workflow)**

There is a bit of a trade-off between fast local frame reads and large jumps in the video (e.g. dragging the slider around quickly). 
I tried to strike a balance between the two while keeping memory usage in mind, let me know how it performs on your video if possible.

Best,
Cyril

**Comment by ChaoqunYinNeuroscience** [comment]
Thanks for the update. I'll try it as soon as possible.!

**Comment by C-Achard** [comment]
Fixed in [napari-deeplabcut v0.3.1.1](https://github.com/DeepLabCut/napari-deeplabcut/releases/tag/v0.3.1.1). Thanks for the report!
